### PR TITLE
fixed extra dataloader bug

### LINKED ISSUE #1181: Additional dataloader created and discarded when training with reload_dataloaders_every_epoch
## 🐛 Bug

I am training with reload_dataloaders_every_epoch and I've noticed it instantiates an extra DataLoader before training for which nothing is run. This is an issue for me as I am training with chunks that get loaded every epoch and it is messing with the order I load them in especially if I reload a checkpoint; it would be an issue for people that generate a new dataset every epoch as they waste computation. The tqdm bar also keeps the information of the first, discarded DataLoader (in the screenshot, the number of iterations is the same for both whereas they should be different sizes)

image

### To Reproduce

Run the code sample below, which runs for one epoch and displays a message every time a DataLoader is created.

A DataLoader gets instantiated a first time line 286 in `training_loop.py` outside of the epoch loop (that's the usual time it gets instantiated when not reloading every epoch. Then when using `reload_dataloaders_every_epoch` another one is created at the start of every epoch line 386, inside the loop, so for the first epoch there's an extra one.

#### Code sample

```
import torch
import pytorch_lightning as pl
from torch.utils.data import DataLoader, Dataset
from time import sleep

class MinimalDataset(Dataset):

    def __init__(self, index):
        self.data = torch.Tensor(64 * index, 1024)

    def __getitem__(self, item):
        return self.data[item]

    def __len__(self):
        return len(self.data)

class MinimalModule(pl.LightningModule):

    def __init__(self):
        super(MinimalModule, self).__init__()
        self.nn = torch.nn.Linear(1024, 1)
        self.current_index = 0

    def forward(self, batch):
        return self.nn(batch)

    def training_step(self, batch, batch_idx):
        sleep(0.1)
        loss = self.nn(batch)[0]
        return {'loss': loss}

    def validation_step(self, batch, batch_idx):
        loss = self.nn(batch)[0]
        return {'val_loss': loss}

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.01)

    def train_dataloader(self):
        # REQUIRED
        self.current_index += 1
        print(f"initializing DataLoader n{self.current_index}")
        data_loader = DataLoader(MinimalDataset(self.current_index))
        return data_loader
    
model = MinimalModule()
trainer = pl.Trainer(reload_dataloaders_every_epoch=True, num_sanity_val_steps=0, val_check_interval=8, max_epochs=1)

trainer.fit(model)
```

### Expected behavior

Only one dataloader should be created; two are. The tqdm bar should show 128 iterations as that is the dataset size the second time; but it shows 64 instead (I added the sleep(0.1) to leave time to observe that)

### Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collectepoch_end
GPU models and configuration: GPU 0: GeForce RTX 2070 with Max-Q Design
Nvidia driver version: 435.21
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] pytorch-lightning==0.7.1
[pip3] torch==1.4.0
[pip3] torchvision==0.4.2
[conda] Could not collect

### PR DESCRIPTION
**Pull request by TevenLeScao**
## What does this PR do?
Fixes #1181 (issue).

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -304,6 +306,10 @@ def train(self):
         try:
             # run all epochs
             for epoch in range(self.current_epoch, self.max_epochs):
+
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]:
```suggestion
```

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -288,7 +288,9 @@ def train(self):
         model = self.get_model()
 
         # load data
-        self.reset_train_dataloader(model)
+        # if reload_dataloaders_every_epoch, this is moved to the epoch loop
+        if not self.reload_dataloaders_every_epoch:
+            self.reset_train_dataloader(model)
         self.reset_val_dataloader(model)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py]: this should be aslo in the iff block, right?

**Comment by ethanwharris** [comment]
This looks good, it's not clear how the `reload_dataloaders_every_epoch' argument is supposed to work for val and test since we often don't do an epoch of those. But this fix is good for the train dataloader - perhaps the argument name should be changed to `reset_train_dataloader_every_epoch'? (and deprecate the old argument)

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (TevenLeScao) [reply pytorch_lightning/trainer/training_loop.py]: Do you mean reset_val_dataloader ? If so, in the original code it wasn't reloaded every epoch so I kept the same behaviour.

**Comment by TevenLeScao** [comment]
I've added a note in CHANGELOG.md

File: CHANGELOG.md [hunk 3]
```diff
@@ -31,6 +31,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fixed bug related to type cheking of `ReduceLROnPlateau` lr schedulers([#1114](https://github.com/PyTorchLightning/pytorch-lightning/issues/1114))
 - Fixed a bug to ensure lightning checkpoints to be backward compatible ([#1132](https://github.com/PyTorchLightning/pytorch-lightning/pull/1132))
+- Fixed a bug that created an extra dataloader with reload_dataloaders_every_epoch active ([#1181](https://github.com/PyTorchLightning/pytorch-lightning/issues/1181)
```
Comment (Borda) [inline CHANGELOG.md]:
```suggestion
- Fixed a bug that created an extra dataloader with active `reload_dataloaders_every_epoch` ([#1181](https://github.com/PyTorchLightning/pytorch-lightning/issues/1181)
```

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (tullie) [inline pytorch_lightning/trainer/training_loop.py]: Will the same problem happen for val_dataloader as well?

Comment (TevenLeScao) [reply pytorch_lightning/trainer/training_loop.py]: val_dataloader doesn't get re-initialized every epoch with reload_dataloaders_every_epoch, the name is a bit misleading as ethanwarris mentioned (and I'm not sure it should)

Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: that was my question above... :]

Comment (williamFalcon) [reply pytorch_lightning/trainer/training_loop.py]: good point @ethanwharris since val dataloader does reset every epoch anyhow.

**Comment by TevenLeScao** [comment]
Sorry, I'm not sure what the next steps are at the moment - @williamFalcon , do you want to follow @ethanwharris 's suggestion of changing `reset_dataloader_every_epoch` to `reset_train_dataloader_every_epoch` ? I'm not sure I understand your last message.

**Comment by williamFalcon** [comment]
@TevenLeScao yes, let's rename to ```reset_train_dataloader_every_epoch```

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 4]
```diff
@@ -338,14 +338,14 @@ def run_evaluation(self, test_mode: bool = False):
 
         # select dataloaders
         if test_mode:
-            if self.reload_dataloaders_every_epoch or self.test_dataloaders is None:
+            if self.reload_train_dataloader_every_epoch or self.test_dataloaders is None:
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/evaluation_loop.py]: this is wrong no? don’t we want to reset the test and val dataloaders with every call to evaluate?

Comment (TevenLeScao) [reply pytorch_lightning/trainer/evaluation_loop.py]: I don't think we want to ever reset the val dataloader as that would defeat the purpose of tracking progress with a validation set; but in any case I believe I haven't changed any of the functionality as the validation set wasn't being reset before.

Comment (TevenLeScao) [reply pytorch_lightning/trainer/evaluation_loop.py]: If we want to keep that, maybe the compromise is to keep the name `reload_dataloaders_every_epoch`, and consider that this reloads the train dataloader every training epoch in `training_loop`, and the val/test dataloaders at every evaluation in `evaluation_loop`. I can just revert the last changes in this case, and the previously-approved PR should be good to go. Sorry if I'm reinventing the wheel here !

Comment (ethanwharris) [reply pytorch_lightning/trainer/evaluation_loop.py]: I thought the point was just that the reload_train_dataloader_every_epoch was doing stuff with test and val when it wasn't needed - not to revert the change?

Comment (TevenLeScao) [reply pytorch_lightning/trainer/evaluation_loop.py]: I'm actually not sure anymore we're on the same page here ! But in any case I think it's better to have `reload_dataloaders_every_epoch` stand for everything, as it keeps previous functionality and doesn't split it into an argument for everything (ie train val and test)

Comment (williamFalcon) [reply pytorch_lightning/trainer/evaluation_loop.py]: evaluate runs the test and val loop.

Maybe i'm missing something, but if the user wants to reload the validation and test datasets every time evaluation is checked that should be allowed no?

Comment (ethanwharris) [reply pytorch_lightning/trainer/evaluation_loop.py]: Haha, I'm confused. Originally tbe point was made that `reload_dataloaders_every_epoch` didn't previously reload the test and val dataloaders, only train. So then I suggested the name change, but it turns out we do reload them? Anyway, let's make it so that `reload_dataloaders_every_epoch` does what it says on the tin and applies to val and test aswell, in which case I think the above comment from @tullie still applies. Can then revisit it if someone finds a use case where that doesn't work for them :)

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py]: @TevenLeScao pls ^^

Comment (TevenLeScao) [reply pytorch_lightning/trainer/training_loop.py]: Yeah it's in the latest commit ! Passing CI atm

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/trainer/evaluation_loop.py]: @ethanwharris is this also solved?

Comment (TevenLeScao) [reply pytorch_lightning/trainer/evaluation_loop.py]: It should be fixed now, but the CI says some checks were cancelled and I'm not sure why :/

**Comment by TevenLeScao** [comment]
Should be good to go now !

**Comment by Borda** [comment]
@williamFalcon ^^

**Comment by TevenLeScao** [comment]
So is there anything left to do here before merging ?
